src/vic3save/src/stats.rs

```diff
@@ -0,0 +1,192 @@
+use crate::Vic3Date;
+use serde::Deserialize;
+use std::collections::HashMap;
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct VicStatsChannel {
+    // Date of last sample
+    pub date: Vic3Date,
+    //  next free index (number of samples + 1)
+    pub index: u32,
+    // Value
+    pub values: Vec<f32>,
+}
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct Vic3CountryStats {
+    // Sample rate of ticks (sample rate of 28, works to 1 week)
+    pub sample_rate: u32,
+    // Not sure, double the index?
+    pub count: u32,
+    pub channels: HashMap<u32, VicStatsChannel>,
+}
+
+#[derive(Debug)]
+pub struct Vic3CountryStatsIter<'a> {
+    stats: &'a Vic3CountryStats,
+    index: usize,
+}
+impl<'a> Iterator for Vic3CountryStatsIter<'a> {
+    type Item = (Vic3Date, f32);
+    fn next(&mut self) -> Option<Self::Item> {
+        let start_date: Vic3Date = Vic3Date::from_ymdh(1836, 1, 1, 0);
+        let channel = &self.stats.channels[&0];
+        if self.index + 1 >= channel.index.try_into().unwrap() {
+            return None;
+        }
+        let elem = channel.values[self.index];
+        self.index += 1;
+        let days_per_index = self.stats.sample_rate / 4;
+
+        let new_date = start_date.add_days((self.index as u32 * days_per_index) as i32);
+        return Some((new_date, elem));
+    }
+}
+
+#[derive(Debug)]
+pub struct Vic3CountryStatsAllignedIter<T1, T2> {
+    a: T1,
+    b: T2,
+}
+impl<A, B, T1, T2> Iterator for Vic3CountryStatsAllignedIter<T1, T2>
+where
+    T1: Iterator<Item = (Vic3Date, A)>,
+    T2: Iterator<Item = (Vic3Date, B)>,
+{
+    type Item = (Vic3Date, (A, B));
+    fn next(&mut self) -> Option<Self::Item> {
+        match (self.a.next(), self.b.next()) {
+            (Some((date_a, x)), Some((date_b, y))) if date_a == date_b => Some((date_a, (x, y))),
+            (Some((date_a, x)), Some((date_b, y))) if date_a < date_b => {
+                let mut prev_date = date_a;
+                let mut prev_x = x;
+                loop {
+                    let Some((date_next, x1)) = self.a.next() else {
+                        return Some((prev_date, (prev_x, y)));
+                    };
+                    if date_next == date_b {
+                        return Some((date_b, (x1, y)));
+                    }
+
+                    if date_next > date_b {
+                        panic!("todo: missing case where sample rate of a isn't a multiple of b (such as 3 and 5)");
```

Comment:
This seems like certain user input could cause a panic. Are you able to code up a workaround? Like maybe panicing in debug, but returning None otherwise.

Reply:
I changed this to just return date_next and print a warning. I'm pretty sure this doesn't happen right now in the saves (atleast the parts parsed). So it would be nice to be alerted if this pops up, but I agree that panic is too strong.